dhis-2/dhis-services/dhis-service-event-hook/src/main/java/org/hisp/dhis/eventhook/EventHookListener.java

```diff
@@ -0,0 +1,161 @@
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class EventHookListener
+{
+    private final ObjectMapper objectMapper;
+
+    private final FieldFilterService fieldFilterService;
+
+    private EventHookContext eventHookContext = EventHookContext.builder().build();
+
+    private final EventHookService eventHookService;
+
+    @Async( "eventHookTaskExecutor" )
+    @TransactionalEventListener( classes = Event.class, phase = TransactionPhase.AFTER_COMMIT, fallbackExecution = true )
+    public void eventListener( Event event )
+        throws JsonProcessingException
+    {
+        for ( EventHook eventHook : eventHookContext.getEventHooks() )
+        {
+            if ( event.getPath().startsWith( eventHook.getSource().getPath() ) )
+            {
+                if ( !eventHookContext.hasTarget( eventHook.getUid() ) )
+                {
+                    continue;
+                }
+
+                if ( event.getObject() instanceof Collection )
+                {
+                    List<ObjectNode> objects = new ArrayList<>();
+
+                    ((Collection<?>) event.getObject()).forEach( object -> {
+                        ObjectNode objectNode = fieldFilterService.toObjectNode( event.getObject(),
+                            List.of( eventHook.getSource().getFields() ) );
+                        objects.add( objectNode );
+                    } );
+
+                    event.setObject( objects );
+                }
+                else
+                {
+                    ObjectNode objectNode = fieldFilterService.toObjectNode( event.getObject(),
+                        List.of( eventHook.getSource().getFields() ) );
+                    event.setObject( objectNode );
+                }
+
+                String payload = objectMapper.writeValueAsString( event );
+
+                List<Handler> handlers = eventHookContext.getTarget( eventHook.getUid() );
+
+                for ( Handler handler : handlers )
+                {
+                    handler.run( eventHook, event, payload );
+                }
+            }
+        }
+    }
+
+    @PostConstruct
+    @EventListener( ReloadEventHookListener.class )
+    public void reload()
+    {
+        eventHookContext.closeTargets();
+
+        List<EventHook> eventHooks = eventHookService.getAll();
+        Map<String, List<Handler>> targets = new HashMap<>();
+
+        for ( EventHook eh : eventHooks )
+        {
+            if ( !eh.isEnabled() )
+            {
+                continue;
+            }
+
+            targets.put( eh.getUid(), new ArrayList<>() );
+
+            for ( Target target : eh.getTargets() )
+            {
+                if ( "webhook".equals( target.getType() ) )
```

Comment:
Use an enum for the various target types? Would be nice to have a central place to store target types I would think.

Reply:
Switch to using constants for json sub-type types.